Fix & optimize open device registration test

fffrog opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124712

Fixes #100152

1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Fix wrong backend meta registry mechanism when compiling with clang++( compiling with g++ work well)(introduced by static variable in inline function)
3. Refactor the tests and make it more flexible
4. Disable the two tests temporarily
    - test_open_device_storage_pin_memory
    - test_compile_autograd_function_aliasing

fffrog committed:
Fix & optimze open device registration test.

1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily and will fix it later
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback
     - test_compile_autograd_function_aliasing

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily and will fix it later
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback
     - test_compile_autograd_function_aliasing

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily and will fix it later
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback
     - test_compile_autograd_function_aliasing

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily and will fix it later
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback
     - test_compile_autograd_function_aliasing

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily and will fix it later
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback
     - test_compile_autograd_function_aliasing

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily and will fix it later
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback
     - test_compile_autograd_function_aliasing

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily and will fix it later
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback
     - test_compile_autograd_function_aliasing

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily and will fix it later
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback
     - test_compile_autograd_function_aliasing

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily and will fix it later
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback
     - test_compile_autograd_function_aliasing

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily and will fix it later
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback
     - test_compile_autograd_function_aliasing

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily and will fix it later
     - test_open_device_faketensor
     - test_open_device_scalar_type_fallback
     - test_compile_autograd_function_aliasing

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Refactor the tests and make it more flexible
3. Disable the two tests temporarily
    - test_open_device_storage_pin_memory
    - test_compile_autograd_function_aliasing
    - test_open_device_scalar_type_fallback

[ghstack-poisoned]

albanD commented on test/test_cpp_extensions_open_device_registration.py:
```diff
@@ -28,30 +29,38 @@ def remove_build_path():
     def current_device():
         return 0
 
+    # create a new module to fake torch.foo dynamicly
```
```suggestion
    # create a new module to fake torch.foo dynamically
```

FFFrog replied:
Done

albanD commented on test/test_cpp_extensions_open_device_registration.py:
```diff
@@ -86,491 +99,462 @@ def setUpClass(cls):
+        self.assertTrue(z_cpu.is_cpu)
+        self.assertFalse(z.is_cpu)
+        self.assertTrue(z.device == device)
+        self.assertEqual(z, z_cpu)
```
nit: this assertEqual is going to re-do the copy of z to the same device as z_cpu so we're only checking that the cross device copy is consistent. It's fine but I would update the comment above.

FFFrog replied:
Got it

albanD commented on test/test_cpp_extensions_open_device_registration.py:
```diff
@@ -86,491 +99,462 @@ def setUpClass(cls):
+        with self.assertRaisesRegex(RuntimeError, "The runtime module of"):
+            torch._register_device_module("foo", generate_faked_module())
+
+        # backend name can only rename once
```
Can we rename it to the same name multiple times?

FFFrog replied:
Yes, Rename it to the same name is allowed, i will add another test to check it.

albanD reviewed (commented):
SGTM!
Only small nits!

fffrog committed:
Update on "Fix & optimze open device registration test."


Fixes #100152

1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Fix wrong backend meta registry mechanism when compiling with clang++( compiling with g++ work well)(introduced by static variable in inline function)
3. Refactor the tests and make it more flexible
4. Disable the three tests temporarily
    - test_open_device_storage_pin_memory
    - test_compile_autograd_function_aliasing
    - test_open_device_scalar_type_fallback

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


Fixes #100152

1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Fix wrong backend meta registry mechanism when compiling with clang++( compiling with g++ work well)(introduced by static variable in inline function)
3. Refactor the tests and make it more flexible
4. Disable the three tests temporarily
    - test_open_device_storage_pin_memory
    - test_compile_autograd_function_aliasing
    - test_open_device_scalar_type_fallback

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


Fixes #100152

1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Fix wrong backend meta registry mechanism when compiling with clang++( compiling with g++ work well)(introduced by static variable in inline function)
3. Refactor the tests and make it more flexible
4. Disable the two tests temporarily
    - test_open_device_storage_pin_memory
    - test_compile_autograd_function_aliasing

[ghstack-poisoned]

fffrog committed:
Update on "Fix & optimze open device registration test."


Fixes #100152

1. Fix the wrong tests about lazy init for PrivateUse1 named foo
2. Fix wrong backend meta registry mechanism when compiling with clang++( compiling with g++ work well)(introduced by static variable in inline function)
3. Refactor the tests and make it more flexible
4. Disable the two tests temporarily
    - test_open_device_storage_pin_memory
    - test_compile_autograd_function_aliasing

[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge

kit1980 commented:
I'm going to revert this as it causes `ld.lld: error: undefined symbol: torch::jit::GetBackendMetaSerialization()` internally.

@albanD @malfet This probably needs internal build rules updates, D56736319 in case you want to try fixing.

kit1980 commented:
@pytorchbot revert -m "breaking internal builds" -c ghfirst

kit1980 commented:
closing this again to re-attempt revert

kit1980 commented:
@pytorchbot revert -m "breaking internal builds" -c ghfirst

albanD commented:
What kind of rule update does this need? It is only one very specific job that seems to fail?

kit1980 commented:
> What kind of rule update does this need? It is only one very specific job that seems to fail?

Probably new dependency needed to be added to buck? I quickly took a look and couldn't see what is needed.

albanD commented on torch/csrc/jit/serialization/pickler.cpp:
```diff
@@ -803,4 +803,18 @@ bool checkHasValidSetGetState(const std::shared_ptr<c10::ClassType>& cls) {
+std::array<
+    c10::optional<std::pair<BackendMetaPtr, BackendMetaPtr>>,
+    at::COMPILE_TIME_MAX_DEVICE_TYPES>&
+GetBackendMetaSerialization() {
```
Can you move this back to the header file by any chance? That would solve the build issue that is going to painful to investigate haha

FFFrog replied:
Of course.

But there are some tiny differences between gcc and clang in defining global variables in inline function, which will cause the test failure if compiled by clang. Moving the function to .cc is the most easiest way to fix the problem.

So if we want to move the function back, I need sometime to dive into the problem and find out the root cause.

Alternative, I can disable the related test temporarily and re-enable it when fixed.

albanD replied:
Ok,
This is very weird because getWriteableTensorData method above is defined the same way as your updated one but does work...

albanD replied:
I can't think of the problem right now.
Let's reland the rest of the fix with skips for the tests that are indeed broken.
This will at least ensure the rest is properly tested. We can look into the fix for this independently.

FFFrog replied:
Got it.
I will submit a new PR today.

kit1980 commented:
@FFFrog You merged into your ghstack base, not main.
Need to use pytorchbot to properly merge.

FFFrog commented:
> @FFFrog You merged into your ghstack base, not main. Did you intend that?

Sorry, I made a mistake, but it has no impact. Just close this PR. I am submitting a new pr to fix the bug now.
(In fact, I confused ghstack checkout with ghstack land. I need to take some time to study ghstack carefully. )
